engine/services/file_watcher_service.h

```diff
@@ -0,0 +1,239 @@
+    FSEventStreamRelease(event_stream);
+    CFRelease(path_to_watch);
+    CFRelease(mypath);
+  }
+
+#elif defined(_WIN32)
+  void watcherThread() {
+    dir_handle =
+        CreateFileA(watchPath.c_str(), FILE_LIST_DIRECTORY,
+                    FILE_SHARE_READ | FILE_SHARE_DELETE, NULL, OPEN_EXISTING,
+                    FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED, NULL);
+
+    if (dir_handle == INVALID_HANDLE_VALUE) {
+      throw std::runtime_error("Failed to open directory");
+    }
+
+    char buffer[4096];
+    OVERLAPPED overlapped = {0};
+    overlapped.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+    DWORD bytesReturned;
+
+    while (running) {
+      if (!ReadDirectoryChangesW(
+              dir_handle, buffer, sizeof(buffer), TRUE,
+              FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_DIR_NAME,
+              &bytesReturned, &overlapped, NULL)) {
+        break;
+      }
+
+      if (WaitForSingleObject(overlapped.hEvent, INFINITE) != WAIT_OBJECT_0) {
+        break;
+      }
+
+      if (!GetOverlappedResult(dir_handle, &overlapped, &bytesReturned,
+                               FALSE)) {
+        break;
+      }
+
+      FILE_NOTIFY_INFORMATION* event = (FILE_NOTIFY_INFORMATION*)buffer;
+      do {
+        if (event->Action == FILE_ACTION_REMOVED) {
+          std::wstring fileName(event->FileName,
+                                event->FileNameLength / sizeof(wchar_t));
+
+          std::string file_name_str(fileName.begin(), fileName.end());
+          model_service_->ForceIndexingModelList();
+        }
+
+        if (event->NextEntryOffset == 0)
+          break;
+        event = (FILE_NOTIFY_INFORMATION*)((uint8_t*)event +
+                                           event->NextEntryOffset);
+      } while (true);
+
+      ResetEvent(overlapped.hEvent);
+    }
+
+    CloseHandle(overlapped.hEvent);
+    CloseHandle(dir_handle);
+  }
+
+#else  // Linux
+
+  void AddWatch(const std::string& dirPath) {
+    wd = inotify_add_watch(fd, dirPath.c_str(),
+                           IN_DELETE | IN_CREATE | IN_DELETE_SELF);
+    if (wd < 0) {
+      throw std::runtime_error("Failed to add watch on: " + dirPath);
+    }
+    watchDescriptors[wd] = dirPath;
+
+    // Recursively add watches to subdirectories
+    for (const auto& entry :
+         std::filesystem::recursive_directory_iterator(dirPath)) {
+      if (std::filesystem::is_directory(entry)) {
+        AddWatch(entry.path().string());
+      }
+    }
+  }
+
+  void CleanupWatches() {
+    for (const auto& [wd, path] : watchDescriptors) {
+      inotify_rm_watch(fd, wd);
+    }
+    watchDescriptors.clear();
+
+    if (fd >= 0) {
+      close(fd);
+      fd = -1;
+    }
+  }
+
+  void watcherThread() {
+    fd = inotify_init();
+    if (fd < 0) {
+      throw std::runtime_error("Failed to initialize inotify");
+    }
+
+    // Add initial watch on the main directory
+    AddWatch(watchPath);
+
+    char buffer[4096];
+    while (running) {
```

Comment:
Should we put a sleep in the while loop?

Reply:
`read()` is a blocking call. I don't think we should add sleep in there.